api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -300,101 +299,88 @@ private DataSetRequest generateDataSetRequestFromPreviewRequest(
+
+      /* Google appengine has a 60 second timeout, we want to make sure this endpoint completes
+       * before that limit is exceeded, or we get a 500 error with the following type:
+       * com.google.apphosting.runtime.HardDeadlineExceededError
+       * See https://cloud.google.com/appengine/articles/deadlineexceedederrors for details
+       */
+      queryResponse =
+          bigQueryService.executeQuery(
+              bigQueryService.filterBigQueryConfig(queryJobConfiguration),
+              APP_ENGINE_HARD_TIMEOUT_MSEC_MINUS_FIVE_SEC);
+    } catch (Exception ex) {
+      if ((ex.getCause() != null
+          && ex.getCause().getMessage() != null
+          && ex.getCause().getMessage().contains("Read timed out"))) {
+        throw new GatewayTimeoutException(
+            "Timeout while querying the CDR to pull preview information.");
+      } else {
+        throw ex;
       }
-    } while (retry < PREVIEW_RETRY_LIMIT);
-
-    if (retry == PREVIEW_RETRY_LIMIT) {
-      throw new GatewayTimeoutException(
-          "Timeout while querying the CDR to pull preview information.");
     }
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
+            fields -> {
+              valuePreviewList.add(new DataSetPreviewValueList().value(fields.getName()));
+            });
+
+    queryResponse
+        .getValues()
+        .forEach(
+            fieldValueList -> {
+              IntStream.range(0, fieldValueList.size())
+                  .forEach(
+                      columnNumber -> {
+                        try {
+                          valuePreviewList
+                              .get(columnNumber)
+                              .addQueryValueItem(
+                                  fieldValueList.get(columnNumber).getValue().toString());
+                        } catch (NullPointerException ex) {
+                          log.severe(
+                              String.format(
+                                  "Null pointer exception while retriving value for query: Column %s ",
+                                  columnNumber));
+                          valuePreviewList.get(columnNumber).addQueryValueItem("");
+                        }
+                      });
+            });
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
+            fields -> {
+              DataSetPreviewValueList previewValue =
+                  valuePreviewList.stream()
+                      .filter(preview -> preview.getValue().equalsIgnoreCase(fields.getName()))
+                      .findFirst()
+                      .get();
+              if (fields.getType() == LegacySQLTypeName.TIMESTAMP) {
+                List<String> queryValues = new ArrayList<String>();
+                DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
+                previewValue
+                    .getQueryValue()
+                    .forEach(
```

Comment:
I think it would be more readable to make a private helper method to do the stuff inside the foreach and return a string, and then just collect that into a list.

Also, using empty strings makes me nervous, but maybe that's how it has to work.

Reply:
Let me poke at the foreach's, because those might be changing anyway.